File: BM24_二叉树的中序遍历/java_solutions/StackSolution.java
package BM24_二叉树的中序遍历.java_solutions;

import java.util.ArrayList;
import java.util.Stack;

public class StackSolution extends Solution {
    public int[] inorderTraversal (TreeNode root) {
        if(root == null) return new int[0];
        
        ArrayList<Integer> res = new ArrayList<>();
        Stack<Object> stack = new Stack<>(); 

        stack.push(root);
        while (!stack.isEmpty()) {
            Object top = stack.pop();
            if(top instanceof Integer) {
                res.add((Integer) top);
                continue;
            }

            TreeNode topNode = (TreeNode) top;

            if(topNode.right != null) stack.push(topNode.right);
            if(topNode.left == null) {
                res.add(topNode.val);
            } else {
                stack.push(topNode.val);
                stack.push(topNode.left);
            }
        }

        return convertToIntArray(res);
    }
}

File: 94_Binary_Tree_Inorder_Traversal/python_solutions/stack_solution.py
# Definition for a binary tree node.
from typing import List, Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root: return []

        result = []
        stack = []

        while root or stack:
            while root:
                stack.append(root)
                root = root.left
            
            top = stack.pop()
            result.append(top.val)
            root = top.right

        return result